node/bindnode/infer.go

```diff
@@ -87,6 +104,78 @@ func fieldNameFromSchema(name string) string {
 	return strings.Title(name)
 }
 
+var defaultTypeSystem schema.TypeSystem
+
+func init() {
+	defaultTypeSystem.Init()
+
+	defaultTypeSystem.Accumulate(schemaTypeBool)
+	defaultTypeSystem.Accumulate(schemaTypeInt)
+	defaultTypeSystem.Accumulate(schemaTypeFloat)
+	defaultTypeSystem.Accumulate(schemaTypeString)
+	defaultTypeSystem.Accumulate(schemaTypeBytes)
+	defaultTypeSystem.Accumulate(schemaTypeLink)
+}
+
+// TODO: support IPLD maps and unions in inferSchema
+
+// TODO: support bringing your own TypeSystem?
+
+// TODO: we should probably avoid re-spawning the same types if the TypeSystem
+// has them, and test that that works as expected
+
 func inferSchema(typ reflect.Type) schema.Type {
-	panic("TODO")
+	switch typ.Kind() {
+	case reflect.Bool:
+		return schemaTypeBool
+	case reflect.Int64:
+		return schemaTypeInt
+	case reflect.Float64:
+		return schemaTypeFloat
+	case reflect.String:
+		return schemaTypeString
+	case reflect.Struct:
+		fieldsSchema := make([]schema.StructField, typ.NumField())
+		for i := range fieldsSchema {
+			field := typ.Field(i)
+			ftyp := field.Type
+			ftypSchema := inferSchema(ftyp)
+			fieldsSchema[i] = schema.SpawnStructField(
+				field.Name, // TODO: allow configuring the name with tags
+				ftypSchema.Name(),
+
+				// TODO: support nullable/optional with tags
+				false,
+				false,
+			)
+		}
+		name := schema.TypeName(typ.Name())
+		if name == "" {
+			panic("TODO: anonymous composite types")
+		}
+		typSchema := schema.SpawnStruct(name, fieldsSchema, nil)
+		defaultTypeSystem.Accumulate(typSchema)
+		return typSchema
+	case reflect.Slice:
+		if typ.Elem().Kind() == reflect.Uint8 {
+			// Special case for []byte.
+			return schemaTypeBytes
+		}
+
+		etyp := typ.Elem()
+		nullable := false
+		if etyp.Kind() == reflect.Ptr {
+			etyp = etyp.Elem()
+			nullable = true
+		}
+		etypSchema := inferSchema(typ.Elem())
+		name := schema.TypeName(typ.Name())
+		if name == "" {
+			name = "List_" + etypSchema.Name()
```

Comment:
It's going to be common to want to do `SomeField []SomeType`, so I think falling back to `List_SomeType` is reasonable. The alternative is to force the user to write `type List_SomeType []SomeType` on the Go side, but that's going to get pretty annoying.

Reply:
Fwiw, I used "dunder" (*cough* if you will) a la `List__SomeType` for this throughout codegen.  The main reason for this was that saying underscores are a reserved character seems a little aggro to me, and saying double-underscores are reserved does not seem as aggro to me.  (So, roughly the same logic python used, that graphql uses, etc.)

Reply:
I don't feel strongly about this, though I do default to shorter names where possible :) To my mind, this is a fallback/default, not necessarily a reserved name.